src/monitors4codegen/monitor_guided_decoding/openai_gen.py

```diff
@@ -0,0 +1,167 @@
+
+    all_text_bytes: bytes = tokenizer.tokenizer.decode_bytes(all_tokens.tolist())
+    prompt_num_tokens: int = all_tokens.shape[0]
+
+    priority_blacklist: List[int] = []
+
+    while all_tokens.shape[0] < prompt_num_tokens + num_new_tokens:
+        num_toks_to_gen = (prompt_num_tokens + num_new_tokens) - all_tokens.shape[0]
+        
+        blacklisted_ids: List[int] = asyncio.run_coroutine_threadsafe(monitor.maskgen(all_tokens.tolist()), monitor.monitor_file_buffer.lsp.server.loop).result()
+        white_listed_ids: Set[int] = set(tokenizer.all_token_ids) - set(blacklisted_ids+[50256])
+
+        logit_bias = {50256:-100}
+
+        for token_id in priority_blacklist:
+            logit_bias[token_id] = -100
+
+        if len(white_listed_ids) <= (300 - len(logit_bias)):
+            for white_token_id in white_listed_ids:
+                logit_bias[white_token_id] = 100
+        else:
+            for candidate_token in sorted(blacklisted_ids, key=lambda x: tokens_sort_key[x], reverse=True):
+                if len(logit_bias) >= 300:
+                    break
+                if candidate_token in blacklisted_ids:
+                    logit_bias[candidate_token] = -100
+
+        exponential_backoff_wait = 1
+        while True:
+            try:
+                prompt_arg: str = all_text_bytes.decode('utf-8', errors='strict')
+            except UnicodeDecodeError:
+                prompt_arg: List[int] = all_tokens.tolist()
+            
+            try:
+                response = client.completions.create(
+                    model=model.value,
+                    prompt=[prompt_arg],
+                    temperature=temp,
+                    max_tokens=num_toks_to_gen if len(logit_bias) <= 1 else 1,
+                    top_p=top_p,
+                    stop=['.'],
+                    logit_bias=logit_bias,
+                    logprobs=5
+                )
+                break
+            except Exception:
+                time.sleep(exponential_backoff_wait)
+                if exponential_backoff_wait < 64:
+                    exponential_backoff_wait = exponential_backoff_wait*2
+                else:
+                    exponential_backoff_wait = 1
+
+        assert len(response.choices) == 1
+
+        def convert_bytesrep_to_bytes(x: str) -> bytes:
+            if x.startswith('bytes:'):
+                return bytes.fromhex(x.replace('bytes:', '').replace('\\x', ''))
+            else:
+                return x.encode()
+
+        tokens_gen_bytes_ = list(map(convert_bytesrep_to_bytes, response.choices[0].logprobs.tokens))
+        tokens_gen_bytes = []
+        dot_found = False
+        for token_bytes in tokens_gen_bytes_:
+            gen_text += token_bytes
+            all_text_bytes += token_bytes
+            tokens_gen_bytes.append(token_bytes)
+            if b'.' in token_bytes:
+                dot_found = True
+                break
+
+        should_manually_add_dot = None
```

Comment:
Add a comment to explain what is happening here and the reason to manually add a dot.

Reply:
Added in PR: https://github.com/microsoft/monitors4codegen/pull/16